libkmod/libkmod-elf.c

```diff
@@ -682,43 +712,34 @@ static int kmod_elf_get_symbols_symtab(const struct kmod_elf *elf,
 	if (strings[i - 1] != '\0')
 		count++;
 
-	/* sizeof(struct kmod_modversion) * count + size + 1 */
+	/* sizeof(struct kmod_modversion) * count + 1 */
 	if (umulsz_overflow(sizeof(struct kmod_modversion), count, &vec_size) ||
-	    uaddsz_overflow(size, vec_size, &tmp_size) ||
-	    uaddsz_overflow(1, tmp_size, &total_size)) {
+	    uaddsz_overflow(1, vec_size, &total_size)) {
 		return -ENOMEM;
 	}
 
 	*array = a = malloc(total_size);
 	if (*array == NULL)
 		return -errno;
 
-	itr = (char *)(a + count);
 	last = 0;
 	for (i = 0, count = 0; i < size; i++) {
 		if (strings[i] == '\0') {
-			size_t slen = i - last;
 			if (last == i) {
 				last = i + 1;
 				continue;
 			}
 			a[count].crc = 0;
 			a[count].bind = KMOD_SYMBOL_GLOBAL;
-			a[count].symbol = itr;
-			memcpy(itr, strings + last, slen);
-			itr[slen] = '\0';
-			itr += slen + 1;
+			a[count].symbol = strings + last;
 			count++;
 			last = i + 1;
 		}
 	}
 	if (strings[i - 1] != '\0') {
-		size_t slen = i - last;
 		a[count].crc = 0;
 		a[count].bind = KMOD_SYMBOL_GLOBAL;
-		a[count].symbol = itr;
-		memcpy(itr, strings + last, slen);
-		itr[slen] = '\0';
+		a[count].symbol = strings + last;
```

Comment:
This could lead to an unterminated string, after all `strings[i - 1] != '\0'`. According to latest changes to kernel and according to spec, such a section is invalid. We should check if the section ends with '\0' and if it does not, consider the section invalid. We have such a check in `kmod_elf_new` already, anyway

Reply:
Nicely spotted. Do you have any reference to these latest changes and the spec?

Reply:
- https://git.kernel.org/pub/scm/linux/kernel/git/modules/linux.git/commit/?h=modules-next&id=837031e052af32c747906238fb1feb87778e4fe0
- man 5 elf
- https://lkml.org/lkml/2024/10/19/334
- https://lkml.org/lkml/2024/10/19/336

Reply:
Awesome, thank you.